docs/design-documents/20231117-better-processors.md

```diff
@@ -0,0 +1,889 @@
+	// there will be no more calls to any other function. After Teardown returns,
+	// the processor will be discarded.
+	Teardown(context.Context) error
+}
+```
+
+Notice that the processor returns its own specification in the first method. This
+is required so that Conduit can build a registry of all available processors
+and give the user more information about them (see
+[Listing processor plugins](#listing-processor-plugins)).
+
+A few notes about the `Process` method:
+
+- Even though we are not adding support for processing multiple records at the
+  same time, we propose to design the interface in a way that will let us add
+  that functionality in the future without breaking changes. This means that the
+  function will for now be called only with a single record.
+- Notice that the return value is a slice of type `ProcessedRecord`. The purpose
+  behind this is to be able to figure out what happened to each input record. This
+  way we can associate each input record to an output record and know what happened
+  to it (was it filtered out, split, merged, or only transformed). This distinction
+  is important because of acknowledgments (e.g. if a record is split into multiple
+  records, Conduit will acknowledge the original record only once all split records
+  are acknowledged).
+- If an error happens in `Process` it should be stored in the `ProcessedRecord` that
+  was being processed when the error occurred. In that case the processor has two
+  options on how to continue processing the remaining records in the slice.
+  1. If the processor has no side effects and cheap it can process the remaining
+     records and return them normally in the slice. Conduit will figure out what to
+     do with the remaining ones based on the dead-letter queue configuration.
+  2. If the processor has side effects or requires a larger amount of time to process
+     a record, we would like to avoid processing records after an error. The remaining
+     records can be filled with a specific `ProcessedRecord` that marks them as not
+     processed. In that case Conduit will determine if it's required to process those
+     records and pass them to the processor in the next call to `Process`.
+
+### Listing processor plugins
+
+We have to differentiate listing actual processors that are attached to actual pipelines
+and listing the specifications of all available processors. We can draw a parallel to
+listing connectors vs. listing connector plugins. We already have a mechanism to list
+actual processors using
+[`ProcessorService.ListProcessors`](https://github.com/ConduitIO/conduit/blob/df048d93e6240c2c3ec5e52af430c05bdfc7a1a3/proto/api/v1/api.proto#L713),
+we need to add a way to list available processors.
+
+Since we are adding pluggable processors we propose to use the same terminology as with
+connectors and introduce **processor plugins**. This would also give us the ability to
```

Comment:
I really like how much of the connector design we can re-use here!

Reply:
Yes, I think it's better than introducing yet another concept.